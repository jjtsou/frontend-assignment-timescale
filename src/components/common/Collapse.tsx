import { ReactNode, useCallback, useState } from 'react';
import {
  Flex,
  IconButton,
  Text,
  Collapse as ChakraCollapse,
  Button,
} from '@chakra-ui/react';
import { ChevronDownIcon, ChevronRightIcon } from '@chakra-ui/icons';

type CollapseTypes = {
  header: string;
  openByDefault?: boolean;
  onClickHeader?: () => void;
  children: ReactNode;
};

//
// ðŸ”´ Avoid: Adjusting state on prop change in an Effect
// useEffect(() => {
//   setSelection(null);
// }, [items]);
//
// https://beta.reactjs.org/learn/you-might-not-need-an-effect
//
// Better: Adjust the state while rendering
// const [prevItems, setPrevItems] = useState(items);
// if (items !== prevItems) {
//   setPrevItems(items);
//   setSelection(null);
// }
//
// I could use this approach to open / close the collapse when it has no items to show

const Collapse = ({
  header,
  openByDefault = false,
  onClickHeader,
  children,
}: CollapseTypes): JSX.Element => {
  const [isOpen, setIsOpen] = useState(openByDefault);
  const handleToggle = useCallback(() => setIsOpen((isOpen) => !isOpen), []);
  return (
    <Flex direction="column">
      <Flex justifyContent="start" alignItems="center">
        <IconButton
          variant="ghost"
          aria-label="Open collapsible panel"
          icon={isOpen ? <ChevronDownIcon /> : <ChevronRightIcon />}
          size="sm"
          onClick={handleToggle}
          minW={6}
        />
        <Text
          as={Button}
          fontWeight="semibold"
          color="gray.500"
          onClick={onClickHeader}
          cursor={onClickHeader ? 'pointer' : 'initial'}
        >
          {header}
        </Text>
      </Flex>
      <ChakraCollapse in={isOpen} animate animateOpacity>
        <Flex direction="column" pl={10}>
          {children}
        </Flex>
      </ChakraCollapse>
    </Flex>
  );
};

export default Collapse;
