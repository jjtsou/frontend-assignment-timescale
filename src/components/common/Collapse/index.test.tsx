import { render } from '@testing-library/react';
import useEvent from '@testing-library/user-event';
import Collapse from '../Collapse';

describe('Collapse', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <Collapse header="Header">
        <span>Content</span>
      </Collapse>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with the header and content visible when is openByDefault', () => {
    const { asFragment } = render(
      <Collapse header="Header" openByDefault>
        <span>Content</span>
      </Collapse>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onClickHeader when header is clicked', async () => {
    const onClickHeader = jest.fn();
    const { getByText } = render(
      <Collapse header="Header" onClickHeader={onClickHeader}>
        <span>Content</span>
      </Collapse>
    );
    const header = getByText('Header');
    await useEvent.click(header);
    expect(onClickHeader).toHaveBeenCalled();
  });
});
