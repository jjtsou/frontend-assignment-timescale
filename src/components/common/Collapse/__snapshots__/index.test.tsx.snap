// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-j7qwjs"
  >
    <div
      class="css-1m29ivc"
    >
      <button
        aria-label="Open collapsible panel"
        class="chakra-button css-9imh88"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            fill="currentColor"
          />
        </svg>
      </button>
      <p
        class="chakra-text chakra-text css-1efywoh"
      >
        Header
      </p>
    </div>
    <div
      class="chakra-collapse"
      style="overflow: hidden; display: none; opacity: 0; height: 0px;"
    >
      <div
        class="css-11qo4fr"
      >
        <span>
          Content
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Collapse renders with the header and content visible when is openByDefault 1`] = `
<DocumentFragment>
  <div
    class="css-j7qwjs"
  >
    <div
      class="css-1m29ivc"
    >
      <button
        aria-label="Open collapsible panel"
        class="chakra-button css-9imh88"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </button>
      <p
        class="chakra-text chakra-text css-1efywoh"
      >
        Header
      </p>
    </div>
    <div
      class="chakra-collapse"
      style="overflow: hidden; display: block; opacity: 1; height: auto;"
    >
      <div
        class="css-11qo4fr"
      >
        <span>
          Content
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
